cm/master.py

```diff
@@ -1463,6 +1386,111 @@ def _add_data_fs(fs_name=None):
+        elif self.app.config.filesystem_templates: # Legacy behavior - process filesystem_templates directly or snaps.yaml
+            self.process_filesystem_templates(cluster_type, pss, storage_type, self.app.config.filesystem_templates)
+        else:
+            if cluster_type != "Test":
+                log.error("No filesystem templates defined for cluster type: {0}".format(cluster_type))
+
+    # TODO: Remove special case handling for Galaxy filesystem - requires change in UI and initialisation logic
+    def process_filesystem_templates(self, cluster_type, pss, storage_type, filesystem_templates):
+        # Turn those data sources into file systems
+        if filesystem_templates:
+            attached_volumes = self.get_attached_volumes()
+            for fs_template in [s for s in filesystem_templates if 'name' in s]:
+                log.debug("Processing filesystem template: {0}".format(fs_template['name']))
+                fs = Filesystem(self.app, fs_template['name'],
+                                svc_roles=ServiceRole.from_string_array(fs_template.get('roles', None)))
+                # Check if an already attached volume maps to the current filesystem
+                att_vol = self.get_vol_if_fs(attached_volumes, fs_template['name'])
+                if att_vol:
+                    log.debug("{0} file system has volume(s) already attached".format(
+                        fs_template['name']))
+                    fs.add_volume(vol_id=att_vol.id,
+                                  size=att_vol.size, from_snapshot_id=att_vol.snapshot_id)
+                elif 'snap_id' in fs_template:
+                    log.debug("There are no volumes already attached for file system {0}"
+                              .format(fs_template['name']))
+                    size = max(fs_template.get('size', 0), fs_template.get('min_size', 0)) # always default to 0 for snaps
+                    if ServiceRole.GALAXY_DATA in ServiceRole.from_string_array(fs_template.get('roles', None)):
+                        size = pss
+                    fs.add_volume(size=size, from_snapshot_id=fs_template['snap_id'])
+                elif 'type' in fs_template:
+                    # TODO: This obviates the need for type archive. Instead, archive should be a volume "data_source"
+                    if 'archive' == fs_template['type'] and 'archive_url' in fs_template:
+                        log.debug("Creating an archive-based ({0}) file system named '{1}' with storage type: {2} and pss: {3}"
+                                  .format(fs_template.get('archive_url'), fs_template['name'], storage_type, pss))
+                        if storage_type == 'volume':
+                            size = max(fs_template.get('size', 10), fs_template.get('min_size', 0))
+                            if ServiceRole.GALAXY_DATA in ServiceRole.from_string_array(
```

Comment:
The other use is Galaxy Data specific logic. If we can get rid of that too, (through the UI) we can move all this logic into some filesystem related class, so that any filesystem can be created based on a template, and they can be populated from some other template.